### ISSUE TITLE
Syntax Proposal for Multilingual Federated Reactors

### ISSUE THREAD

**Issue by mew2ub** [icyphy/lingua-franca#196]
I've come with some ideas for a syntax for multilingual federated reactors and I'd like to get some feedback.

First a little background. I've tested multilingual federated reactors before, but with a hack. I'll write two versions of the same federated reactor, one in C and one in TypeScript. For the sake of discussion the federated reactor contains two federates, A and B. A sends a message to B. I'll compile the C version and the TypeScript version seperately. Then I'll start the RTI (which is the same for both C and Typescript), the C version of A, and the TypeScript version of B. This is bad for a lot of reasons including: 

- The TypeScript version of A and the C version of B exist only to trick the LF compiler into producing an RTI with the right federate topology. They're just dead code.

- Any change to the federated reactor needs to be reflected in both the TypeScript and C versions.

A lot of these problems could be avoided with the help of dedicated syntax for multilingual federates. The main things I see that need to be accomplished are:

- Designating target properties for each target language in the federate

- Labeling federates with their target.

This leads me to the following as a proposal for the A sends to B example from before.

```
target C {
    timeout: 10 secs
};

target TypeScript {
    timeout: 10 secs
};

C reactor A {
    output message:char*;
    ...
}

TypeScript reactor B {
    input message:Buffer;
    ...
}

federated reactor Distributed at localhost:15044 {
    msg = new A();
    dsp = new B();
    msg.message -> dsp.message after 10 msec;
}
```

Note the duplicate target declarations and the labels "C" and "TypeScript". Does something like this look reasonable?

This way the C reactors can be sent to the C code generator, the TypeScript reactors can be sent to the TypeScript code generator, and the federated reactor (Distributed) can be sent to both generators to establish the federate topology. This last point implies you wouldn't be able to use any target language code (neither C nor TypeScript) in the federated reactor or else you'd get compile/type errors with one of the generators.

**Comment by lhstrh** [comment]
This ties into the discussion of imports, packages, and target properties. We should weigh the desire for multi-lingual federations into the design; IMO, it is too early to discuss concrete syntax for something like this. It is not obvious to me at all, for instance, that we'd want to allow reactors written in different target languages to co-exist in the same file. We really need to address more elementary questions first.

**Comment by edwardalee** [comment]
I'm not sure we actually need much in the way of new syntax.  We could do this:
```
target Multiple {
    timeout: 10 secs
};

import A.lf;
import B.lf;

federated reactor Distributed at localhost:15044 {
    msg = new A();
    dsp = new B();
    msg.message -> dsp.message after 10 msec;
}
```
This file has no target-language code, so a `Multiple` target could check for that an invoke the requisite code generators as needed.

**Comment by mew2ub** [comment]
@lhstrh Yeah, it definitely relates to that discussion. But I'd rather start a branch for a temporary syntax for multilingual reactors now, then wait for the complete resolution of the many issues involving imports. That way I can start working out some of the other interesting issues involved with code generation. And I think concrete examples will make the "same file"/ "different files" discussion more fruitful.

@edwardalee That looks like a very clean way to do it! You'd need to parse `A.lf` and `B.lf`'s target declaration to figure out what generator to use, right? I assume the target properties from `A.lf` and `B.lf` would be discarded which is why there's the new properties for `Multiple`. I'll see if I can make something like that work.

**Comment by edwardalee** [comment]
The MutlipleGenerator could parse the .lf file and then hand off the eCore model to the subtargets. Each subtarget would need to be modified to make “Multiple” an acceptable target, and it would just ignore federates that are not in its language.

**Comment by lhstrh** [comment]
Sounds more like a "None" than a "Multiple" to me (i.e., no target code allowed). It also seems to me that this only makes sense in a federated context, _with the explicit assumption that all federated reactors use the same RTI implementation_ (as is currently the case, but we might want to change this). This solution strikes me as odd and not very extensible. It also overloads the meaning of target for a very specific use case, which I'm apprehensive about.

Perhaps the target of a federated reactor program should instead indicate what RTI to generate (i.e., C, Cpp, Rust, etc.). In this case, all _current_ demos would have to specify `target C` and use imports to point to federates written in different languages. Future demos could use a different RTI. This, to me, seems more descriptive of what's actually happening. This would also obviate this strange construction of mixing C code gen stuff into the `GeneratorBase` class that used to be entirely generic before support for TypeScript federates was merged into master.

**Comment by mew2ub** [comment]
"Multiple" would refer to the targets for reactors instantiated in the federation, not the code written directly in the file. It indicates which code generator should be used. You could create a "C" target file that doesn't directly have any C code because it imports all of its reactors. Under your logic, shouldn't that file's target be "None" too?

What's problematic about always using the same RTI implementation? My understanding is the RTI is always going to be a separate process that communicates with the rest of the federation over a network. Since the RTI doesn't execute target code, its only interaction with the rest of the federation is over the network. The RTI is a black box and only the RTI's ability to support its API matters. The RTI's implementation in C is pretty much irrelevant unless you couldn't run a C program for some reason.

I think we should keep existing target declarations the same. Is there an immediate use-case for changing them?

The reason why the `Federated` reactor can't have target code has nothing to do with the RTI. It's because the `Federated` reactor has to be instantiated on every federate.

**Comment by edwardalee** [comment]
It does seem like having the top-level target be C could work.  The C target would accept federates in other languages and generate a C RTI.  However, it may seem a bit odd to a user who is writing absolutely no C code (e.g., getting Python federates to talk to TypeScript ones) to have to declare the top-level target to be C.

**Comment by lhstrh** [comment]
> "Multiple" would refer to the targets for reactors instantiated in the federation, not the code written directly in the file.

Sure, it would, but that would require changing the definition of what `target` means. I'm not in favor of that.

> It indicates which code generator should be used.

It actually doesn't. The target property of each federate already does that.

> You could create a "C" target file that doesn't directly have any C code because it imports all of its reactors.

Indeed. What this would mean is that the code generator used to generate the RTI would be `CGenerator`. It would allow us to put the C-specific code that now sits in `GeneratorBase` back in `CGenerator` where it belongs.

> Under your logic, shouldn't that file's target be "None" too?

No, "None" is not a language and neither is "Multiple".

> What's problematic about always using the same RTI implementation?

Because we might want multiple. I, for one, would be interested in building a Rust-based RTI, or one in TypeScript. If we cement the RTI written in C as our one-and-only RTI implementation, then this would require a lot of refactoring. That's what I mean by "not very extensible."


> My understanding is the RTI is always going to be a separate process that communicates with the rest of the federation over a network.

It is.

> Since the RTI doesn't execute target code, its only interaction with the rest of the federation is over the network.

That's right. It doesn't execute target code; the RTI coordinates an execution. It is a program generated as part of the compilation of a federated reactor.

> The RTI is a black box and only the RTI's ability to support its API matters. The RTI's implementation in C is pretty much irrelevant unless you couldn't run a C program for some reason.

Specifying which code generator to use to produce an RTI doesn't mean that we stop treating it as a black box during execution. We only give a way to specify the language of its implementation, which the target property would seem an appropriate means for.

> I think we should keep existing target declarations the same. Is there an immediate use-case for changing them?

Yes, it would present an opportunity to address the issues I raised at the time you submitted your pull request and clean up `GeneratorBase`. It would make explicit and modifiable what RTI is used and avoid a hard-coded dependency of every target on C.

> The reason why the Federated reactor can't have target code has nothing to do with the RTI. It's because the Federated reactor has to be instantiated on every federate.

???

**Comment by lhstrh** [comment]
> It does seem like having the top-level target be C could work. The C target would accept federates in other languages and generate a C RTI. However, it may seem a bit odd to a user who is writing absolutely no C code (e.g., getting Python federates to talk to TypeScript ones) to have to declare the top-level target to be C.

I don't find it confusing at all. It describes exactly what it does. I think it's confusing to generate C code as a result of compiling LF code that declares `target TypeScript`.

**Comment by mew2ub** [comment]
>> "Multiple" would refer to the targets for reactors instantiated in the federation, not the code written directly in the file.

> Sure, it would, but that would require changing the definition of what target means. I'm not in favor of that.

I don't understand how that's changing the definition. Currently the wiki says: "A target specification defines the target language and the runtime infrastructure to be used in that target language." A "Multiple" target specification indicates that multiple languages and runtime infrastructures will be produced. A "None" specification is inconsistent with the definition because the compiler is not generating "none" target language code.

>> You could create a "C" target file that doesn't directly have any C code because it imports all of its reactors.

> Indeed. What this would mean is that the code generator used to generate the RTI would be CGenerator. It would allow us to put the C-specific code that now sits in GeneratorBase back in CGenerator where it belongs.

I don't think we're talking about the same thing. I was talking about a (non-federated) program that looked something like this:

```
target ????? {
    timeout: 10 secs
};

import A.lf;
import B.lf;

main reactor X {
    msg = new A();
    dsp = new B();
    msg.message -> dsp.message after 10 msec;
}
```

If A.lf and B.lf were both C target reactors, you wouldn't fill in the ???????? target with "None" would you? Yet, this program is almost identical to the multilingual federate example for which you are proposing "None" as a target.

>> The reason why the Federated reactor can't have target code has nothing to do with the RTI. It's because the Federated reactor has to be instantiated on every federate.

> ???

Yeah, this is the current implementation in both C and TypeScript. An AST transformation removes the connection between federates and replaces it with actions and reactions in the federated reactor. Then the target code generator creates for each federate a separate target language file in which the federated reactor contains that federate. If you had a C federate and a TS federate, each of the target language files would wind up with an instance of the federated reactor. So it wouldn't work in one of the targets if the federated reactor contained anything target specific.

**Comment by lhstrh** [comment]
> > > "Multiple" would refer to the targets for reactors instantiated in the federation, not the code written directly in the file.
> 
> > Sure, it would, but that would require changing the definition of what target means. I'm not in favor of that.
> 
> I don't understand how that's changing the definition. Currently the wiki says: "A target specification defines the target language and the runtime infrastructure to be used in that target language." A "Multiple" target specification indicates that multiple languages and runtime infrastructures will be produced. A "None" specification is inconsistent with the definition because the compiler is not generating "none" target language code.

I've never written a program in None, and I think "Multiple" would be a strange name for a particular runtime. I'm not proposing a `None` target. I don't understand why you keep bringing this up.

> > > You could create a "C" target file that doesn't directly have any C code because it imports all of its reactors.
> 
> > Indeed. What this would mean is that the code generator used to generate the RTI would be CGenerator. It would allow us to put the C-specific code that now sits in GeneratorBase back in CGenerator where it belongs.
> 
> I don't think we're talking about the same thing. I was talking about a (non-federated) program that looked something like this:
> 
> ```
> target ????? {
>     timeout: 10 secs
> };
> 
> import A.lf;
> import B.lf;
> 
> main reactor X {
>     msg = new A();
>     dsp = new B();
>     msg.message -> dsp.message after 10 msec;
> }
> ```
>

You're right, we're not talking about the same thing. What you appear to be talking about -- an unfederated LF program with reactors written in different languages -- isn't currently possible, nor do I think that this should be a goal.

> If A.lf and B.lf were both C target reactors, you wouldn't fill in the ???????? target with "None" would you? Yet, this program is almost identical to the multilingual federate example for which you are proposing "None" as a target.

Same comment as before -- I'm not proposing a target called `None`. I'm saying we should interpret the target to specify the language in which the reactors in a file are written, not only for ordinary reactor programs but also for federated ones.

> > > The reason why the Federated reactor can't have target code has nothing to do with the RTI. It's because the Federated reactor has to be instantiated on every federate.
> 
> > ???
> 
> Yeah, this is the current implementation in both C and TypeScript. An AST transformation removes the connection between federates and replaces it with actions and reactions in the federated reactor. Then the target code generator creates for each federate a separate target language file in which the federated reactor contains that federate. If you had a C federate and a TS federate, each of the target language files would wind up with an instance of the federated reactor. So it wouldn't work in one of the targets if the federated reactor contained anything target specific.

Your explanation merely shows that this behavior probably isn't the behavior we want (i.e., things break if the federated reactor happens to have reaction code that is not written in the same language as all of its federates). A federated reactor, like any reactor, can have reactions (I just tried this out by modifying one of the C tests in master). I think these reactions should execute on the RTI _only_ (currently they execute on all federates) and should therefore be written in the language specified as the target of the file that has the definition of this federated reactor. This is perfectly consistent with the interpretation of `target` we've used so far (up until the implementation of federated reactors in TypeScript, which breaks with this), and it prevents things from failing unexpectedly when one writes what otherwise would be considered perfectly legal Lingua Franca code.

**Comment by cmnrd** [comment]
I think that the solution Edward proposed is lean, simple and sticks to the principle we have at the moment. I would suggest a small change though: Do not name it `Multiple` but `Federated`. To me, this makes it obvious that this file does not describe a program written in a specific language, but a federated program consisting of multiple binaries. So in my view, it sticks to 
our current 'definition' of `target`.
```
target Federated {
    timeout: 10 secs
    // here we could add more properties to configure the RTI
};

import A.lf;
import B.lf;

federated reactor Distributed at localhost:15044 {
    msg = new A();
    dsp = new B();
    msg.message -> dsp.message after 10 msec;
}
```

The `Federated` target would not be allowed to contain any reactions and only one single top-level reactor. Actually, the `federated` keyword is redundant here, and we could think of using `main` instead. Any imported files would be processed according to their target properties.

**Comment by edwardalee** [comment]
Yes, “Federated” is better.

**Comment by lhstrh** [comment]
I agree that Edward's proposal is simple, but IMO it is _too simple_. Let me explain what I mean by that.

A top-level reactor in a federated program should not be stripped of its ability to define reactions or mutations. _I think this would be a big mistake._ Suppose I want to:
 - do load balancing
 - do central logging
 - respond to stimuli on the RTI's host, such as sensor input or system events such as exceeding a certain CPU or memory utilization
 - ...
I would use reactions and/or mutations for all of those things. And I would define them in the language that is specified as the target. Note that I'm proposing nothing new here, or something more complicated than what Edward is proposing; all I argue that we consistently use the target property to denote what we (or I, at least) had originally meant it to denote, namely the target language in which reactors are implemented. As such, it also indicates with code generator to use. This is a pretty fundamental notion and I think we'd need a darn good reason to deviate from that interpretation.

Even the proposed `Federated` is redundant in your proposal; it is already implied by the `coordination` target property.
Note that under this interpretation that a federated reactor could not have any reactions of mutations, I would argue it isn't actually a reactor and therefore shouldn't be called a `federated reactor` but a `federation`. I don't subscribe to this view, however, and I propose we keep using the `federated` modifier to indicate a top-level federated reactor. In fact, I should emphasize that I'm proposing no syntax changes at all.

**Comment by edwardalee** [comment]
It's not clear to me what you *are* proposing. Are you proposing that the federation specify `target C` and that CGenerator support instantiating reactors in other languages?  We can certainly do that, but then the C target is the only target supporting multi-lingual federations, and if any other target wants to also support them, it will have to do so from scratch.

**Comment by lhstrh** [comment]
I'm proposing this:

1. A federated reactor, like any other reactor, can have reactions.
2. The reactions of a federated reactor execute on the RTI, and on the RTI _only_.
3. The language in which reactions of a federated reactor are written is the same language in which the RTI is implemented.

Currently, the only RTI implementation we have is in C. So, indeed, any federated reactor program would have to declare its target to be `C`. `GeneratorBase` will have to delegate the code generation of each reactor class to the code generator that corresponds to the target property found in the file in which it is defined. This is a general capability, implementable for all targets at once, as opposed to the current implementation which relies on C code being generated in `GeneratorBase`.

**Comment by edwardalee** [comment]
This doesn't look workable to me, unfortunately. Currently, an AST transformation converts a connection between two federates into two reactions in the top-level reactor, one to send data and one to receive it. These reactions will have to be in the target language of the federate, not the target language of the top-level. So this top level reactor has to have reactions in multiple languages. Of course, you could redesign the whole mechanism some other way, but I don't really see the need for that.

I suggest we proceed as follows: Simply augment the C target to be able to instantiate reactors in other target languages. It will need to identify the target language and invoke methods from that target's code generator to get the reactor code for the above two reactors.  This mechanism will not really be C-specific, but if you really don't want to identify it as such, I can live with that. Only the C target will support multi-lingual federations.

**Comment by edwardalee** [comment]
Oh, also, reactions at the top level do not run in the RTI. No reactions run in the RTI. A reaction at the top level runs in the federate it interacts with (the one that triggers it or the one that receives data from it). Reactions that interact with more than one federate will not currently work at the top level (in fact, I don't know how to make them work).

**Comment by lhstrh** [comment]
> This doesn't look workable to me, unfortunately. Currently, an AST transformation converts a connection between two federates into two reactions in the top-level reactor, one to send data and one to receive it. These reactions will have to be in the target language of the federate, not the target language of the top-level. So this top level reactor has to have reactions in multiple languages. Of course, you could redesign the whole mechanism some other way, but I don't really see the need for that.

I don't really agree with this description. There really is no "one" top-level reactor; there is one for each federate, and each of them only needs to know the topology of the connections in order to synthesize the appropriate send/receive mechanisms as per said AST transformation. I don't see anything standing in the way of the solution I outlined, frankly.

> I suggest we proceed as follows: Simply augment the C target to be able to instantiate reactors in other target languages.
It will need to identify the target language and invoke methods from that target's code generator to get the reactor code for the above two reactors. This mechanism will not really be C-specific, but if you really don't want to identify it as such, I can live with that. Only the C target will support multi-lingual federations.

I don't see how this is better or easier than what I describe above. Perhaps we should do a Zoom call so I can clarify what I mean.

On where top-level reactions should execute: I really think they shouldn't execute on any of the federates. The fact that this makes it impossible for a reaction to be triggered by outputs from multiple contained reactors only underlines the conceptual problem with this. We'd be straying away from the essence of reactors by doing this. If we go this route -- which I suggest we don't -- a federated reactor shouldn't be called a "reactor" but a "federation" because it doesn't behave like a reactor.

**Comment by mew2ub** [comment]
@lhstrh I don't understand what your proposal/solution is because the three items you listed above

>1) A federated reactor, like any other reactor, can have reactions.
>2) The reactions of a federated reactor execute on the RTI, and on the RTI only.
>3) The language in which reactions of a federated reactor are written is the same language in which the RTI is implemented.

refer to executing reactors on the RTI. The RTI _isn't_ a reactor, it's runtime infrastructure that relays messages between federates and controls the advancement of time in the federation. You can't give the RTI reactions because it's not a reactor.
Can you rephrase?

> On where top-level reactions should execute: I really think they shouldn't execute on any of the federates. The fact that this makes it impossible for a reaction to be triggered by outputs from multiple contained reactors only underlines the conceptual problem with this. 

I don't think it's impossible. If you want a reaction to execute in response to output from multiple federates, just create a federate that receives those outputs and give it a reaction to them.

**Comment by edwardalee** [comment]
Marten and I just had a long discussion about this issue because it became clear that we had divergent views about how federations should work.  The differences are interesting (see below if you are curious), but the bottom line is that the current code generation mechanism for federations will need to change significantly. This isn't going to happen right away, and when it does, much of what we have realized will also have to change.

That said, I feel there is still great value in putting together an existence proof for a multilingual federation, and I believe Matt is extremely close to being able to pull this off using the current architecture. I suggest that he do this in the C target, so the top level of federation will have `target C`, but will be able to import and instantiate TypeScript federates.  @mew2ub : Let me know if you need help with this. The first step is to figure out the hooks to hand a parsed AST off to the TypeScript code generator from the CGenerator when the federate's target is TS.

I will go ahead and close this issue since we have a near term plan and a separate long term plan (the latter of which has many details to work out).

So, back to the divergent views:

Marten's position is a declaration at the top level of `target C` means that the RTI is implemented in C and that all top-level reactions are implemented in C as part of the RTI. The federates can still be in other languages, but `target C` is a statement about the top-level "federate".

In the current architecture, if Matt can pull this off, there will be top-level reactions in multiple languages, which is why `target Federated` made sense to me.  Once the rearchitecting is done, there will be no need for top-level reactions for communication, so this issue goes away. All top-level reactions will be user-written code and will be in just one language (initially, only C).

I had in my head that the top-level (federated) reactor exists in *all* federates. Marten's view is that it exists only once, most naturally as part of the RTI.  Marten has convinced me that his view makes more sense because in my view, it is not clear what reactions at the top level mean, and reactions that interact with more than one federate will be particularly difficult to make work (or even to assign meaning to them).  The rearchitecting that will need to happen includes at least these steps:

1. Replacing the current AST transformation handling communication between federates. Instead of generating reactions, this will have to splice in reactors. This means dealing with the headache of target-specific type systems (as we did with the reactors that implement `after`), so it's a nontrivial job.

2. Replacing communication between a top-level reaction and a federate, which currently occurs only within the federate, so that it becomes communication between the RTI and the federate.

3. Identify the limitations on what can be supported in top-level reactors. The paths of communication between the federates and the RTI will become more complex and will affect the mechanisms we use to advance logical time in ways that, at this time, I can only guess.

**Comment by edwardalee** [comment]
(Just seeing Matt's latest comment) Adding to this: Marten was talking about what he sees as a what should be, not what is currently implemented.  So to follow Marten's advice, you would have to first redesign how federations are implemented. This is not possible in the time you have left, which is why I am suggesting you build on the current design, not on a hoped-for future. Marten has convinced me that his hoped-for future is better than the present, but it's a sizable redesign. If you can create a multilingual regression test or two, these will actually help us later do the redesign.

**Comment by mew2ub** [comment]
Thanks for clarifying, it helps a lot to understand that Marten wasn't talking about our current implementation. I'm still not entirely clear on Marten's proposal for the future. Will each of our hypothetical future RTI's be a reactor? Because that's a requirement for giving it reactions, right? I have a lot of other questions too, but I suppose they don't matter given my timeframe.

@edwardalee I'll let you know if I get stuck finding the right hooks.

**Comment by edwardalee** [comment]
I too have lots of questions about his proposal. I don't think those questions will get answered until the design gets done.
